diff --git a/node_modules/typescript/lib/lib.dom.d.ts b/node_modules/typescript/lib/lib.dom.d.ts
index 0f98d76..8e65d13 100644
--- a/node_modules/typescript/lib/lib.dom.d.ts
+++ b/node_modules/typescript/lib/lib.dom.d.ts
@@ -16663,10 +16663,16 @@ interface Window extends EventTarget, AnimationFrameProvider, GlobalEventHandler
     [index: number]: Window;
 }
 
+declare namespace WindowNS {
+    declare var Option: {
+        new(text?: string, value?: string, defaultSelected?: boolean, selected?: boolean): HTMLOptionElement;
+    };    
+}
+
 declare var Window: {
     prototype: Window;
     new(): Window;
-};
+} & typeof WindowNS;
 
 interface WindowEventHandlersEventMap {
     "afterprint": Event;
@@ -17613,9 +17619,6 @@ declare var Audio: {
 declare var Image: {
     new(width?: number, height?: number): HTMLImageElement;
 };
-declare var Option: {
-    new(text?: string, value?: string, defaultSelected?: boolean, selected?: boolean): HTMLOptionElement;
-};
 /** @deprecated This is a legacy alias of `navigator`. */
 declare var clientInformation: Navigator;
 /** Returns true if the window has been closed, false otherwise. */
