diff --git a/node_modules/typescript/lib/tsserver.js b/node_modules/typescript/lib/tsserver.js
index f4b7750..ed34f92 100644
--- a/node_modules/typescript/lib/tsserver.js
+++ b/node_modules/typescript/lib/tsserver.js
@@ -13602,7 +13602,7 @@ var ts;
 (function (ts) {
     ts.resolvingEmptyArray = [];
     ts.externalHelpersModuleNameText = "tslib";
-    ts.defaultMaximumTruncationLength = 160;
+    ts.defaultMaximumTruncationLength = 800;
     ts.noTruncationMaximumTruncationLength = 1000000;
     function getDeclarationOfKind(symbol, kind) {
         var declarations = symbol.declarations;
@@ -165589,7 +165589,7 @@ var ts;
                     return;
                 }
                 // No need to analyze lib.d.ts
-                var fileNamesInProject = fileNames.filter(function (value) { return !ts.stringContains(value, "lib.d.ts"); }); // TODO: GH#18217
+                var fileNamesInProject = fileNames.filter(function (value) { return !ts.stringContains(value, "lib.d.ts") && !ts.stringContains(value, "node_modules"); }); // TODO: GH#18217
                 if (fileNamesInProject.length === 0) {
                     return;
                 }
diff --git a/node_modules/typescript/lib/tsserverlibrary.js b/node_modules/typescript/lib/tsserverlibrary.js
index 1612f5b..f56e5e1 100644
--- a/node_modules/typescript/lib/tsserverlibrary.js
+++ b/node_modules/typescript/lib/tsserverlibrary.js
@@ -13796,7 +13796,7 @@ var ts;
 (function (ts) {
     ts.resolvingEmptyArray = [];
     ts.externalHelpersModuleNameText = "tslib";
-    ts.defaultMaximumTruncationLength = 160;
+    ts.defaultMaximumTruncationLength = 800;
     ts.noTruncationMaximumTruncationLength = 1000000;
     function getDeclarationOfKind(symbol, kind) {
         var declarations = symbol.declarations;
@@ -165783,7 +165783,7 @@ var ts;
                     return;
                 }
                 // No need to analyze lib.d.ts
-                var fileNamesInProject = fileNames.filter(function (value) { return !ts.stringContains(value, "lib.d.ts"); }); // TODO: GH#18217
+                var fileNamesInProject = fileNames.filter(function (value) { return !ts.stringContains(value, "lib.d.ts") && !ts.stringContains(value, "node_modules"); }); // TODO: GH#18217
                 if (fileNamesInProject.length === 0) {
                     return;
                 }
